File: src/main/java/kr/co/bookand/backend/article/service/ArticleService.java
package kr.co.bookand.backend.article.service;

import kr.co.bookand.backend.account.service.AccountService;
import kr.co.bookand.backend.article.domain.Article;
import kr.co.bookand.backend.article.domain.ArticleBookStore;
import kr.co.bookand.backend.article.domain.ArticleCategory;
import kr.co.bookand.backend.article.domain.ArticleTag;
import kr.co.bookand.backend.article.exception.ArticleException;
import kr.co.bookand.backend.article.repository.ArticleBookStoreRepository;
import kr.co.bookand.backend.article.repository.ArticleRepository;

import kr.co.bookand.backend.article.repository.ArticleTagRepository;
import kr.co.bookand.backend.bookmark.domain.BookmarkType;
import kr.co.bookand.backend.bookmark.service.BookmarkService;
import kr.co.bookand.backend.bookstore.domain.BookStore;
import kr.co.bookand.backend.bookstore.repository.BookStoreRepository;
import kr.co.bookand.backend.common.domain.Status;

import kr.co.bookand.backend.common.exception.ErrorCode;
import kr.co.bookand.backend.common.domain.Message;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;


import java.util.ArrayList;
import java.util.List;

import static kr.co.bookand.backend.article.domain.dto.ArticleDto.*;
import static kr.co.bookand.backend.bookstore.domain.dto.BookStoreDto.*;
import static kr.co.bookand.backend.common.domain.dto.PageStateDto.*;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class ArticleService {

    private final ArticleRepository articleRepository;
    private final AccountService accountService;
    private final BookStoreRepository bookStoreRepository;
    private final ArticleBookStoreRepository articleBookStoreRepository;
    private final ArticleTagRepository articleTagRepository;
    private final BookmarkService bookmarkService;

    @Transactional
    public ArticleResponse createArticle(ArticleRequest articleRequest) {
        accountService.isAccountAdmin();
        duplicateArticle(articleRequest.title());
        List<Long> bookStores = articleRequest.bookStoreList();
        List<BookStore> bookStoreList = bookStores.stream()
                .map(bookStoreId -> bookStoreRepository.findById(bookStoreId)
                        .orElseThrow(() -> new ArticleException(ErrorCode.NOT_FOUND_BOOKSTORE, bookStoreId)))
                .toList();
        Article article = articleRequest.toEntity();
        Article saveArticle = articleRepository.save(article);

        List<ArticleTag> tags = articleRequest.tags().stream()
                .map((String id) -> {
                    ArticleTag articleTag = ArticleTag.builder()
                            .tag(id)
                            .article(saveArticle)
                            .build();
                    articleTagRepository.save(articleTag);
                    return articleTag;
                })
                .toList();

        saveArticle.updateArticleTagList(tags);
        return getArticleResponse(bookStoreList, saveArticle);
    }

    private ArticleResponse getArticleResponse(List<BookStore> bookStoreList, Article saveArticle) {
        List<BookStoreSimpleResponse> bookStoreSimpleResponse = new ArrayList<>();

        for (BookStore bookStore : bookStoreList) {
            ArticleBookStore articleBookStore = ArticleBookStore.of(saveArticle, bookStore);
            articleBookStoreRepository.save(articleBookStore);
            bookStore.updateArticleBookStore(articleBookStore);
            saveArticle.addArticleBookStore(articleBookStore);
            bookStoreSimpleResponse.add(BookStoreSimpleResponse.of(bookStore, false));
        }
        boolean bookmark = bookmarkService.isBookmark(saveArticle.getId(), BookmarkType.ARTICLE.toString());
        return ArticleResponse.of(saveArticle, bookStoreSimpleResponse, bookmark);
    }

    public void duplicateArticle(String title) {
        if (articleRepository.existsByTitle(title)) {
            throw new ArticleException(ErrorCode.DUPLICATE_ARTICLE, title);
        }
    }

    // 상세 조회 (APP)
    public ArticleResponse getArticle(Long id) {
        Article article = articleRepository.findById(id)
                .orElseThrow(() -> new ArticleException(ErrorCode.NOT_FOUND_ARTICLE, id));
        List<BookStoreSimpleResponse> bookStoreSimpleResponse = new ArrayList<>();
        article.getArticleBookStoreList()
                .stream().filter(articleBookStore -> articleBookStore.getBookStore().getStatus().equals(Status.VISIBLE))
                .forEach(articleBookStore -> {
            BookStore bookStore = articleBookStore.getBookStore();
            boolean isBookmark = bookmarkService.isBookmark(bookStore.getId(), BookmarkType.ARTICLE.toString());
            bookStoreSimpleResponse.add(BookStoreSimpleResponse.of(bookStore, isBookmark));
        });
        boolean bookmark = bookmarkService.isBookmark(id, BookmarkType.ARTICLE.toString());
        return ArticleResponse.of(article, bookStoreSimpleResponse, bookmark);
    }

    // 전체 조회 (APP)
    public ArticleSimplePageResponse getSimpleArticleList(Pageable pageable) {
        Page<ArticleSimpleResponse> articlePage = articleRepository.findAllByStatus(Status.VISIBLE, pageable)
                .map((Article article) -> ArticleSimpleResponse
                        .of(article, bookmarkService.isBookmark(article.getId(), BookmarkType.ARTICLE.toString())));
        return ArticleSimplePageResponse.of(articlePage);
    }

    // 전체 조회 (WEB)
    public ArticleWebPageResponse getArticleList(Pageable pageable) {
        Page<ArticleWebResponse> articlePage = articleRepository.findAll(pageable)
                .map(ArticleWebResponse::of);
        return ArticleWebPageResponse.of(articlePage);
    }

    // 상세 조회 (APP)
    public ArticleSimpleResponse getSimpleArticle(Long id) {
        Article article = articleRepository.findById(id)
                .orElseThrow(() -> new ArticleException(ErrorCode.NOT_FOUND_ARTICLE, id));
        return ArticleSimpleResponse
                .of(article, bookmarkService.isBookmark(article.getId(), BookmarkType.ARTICLE.toString()));
    }

    @Transactional
    public ArticleResponse updateArticle(Long id, ArticleRequest articleRequest) {
        accountService.isAccountAdmin();
        List<Long> longs = articleRequest.bookStoreList();
        List<BookStore> bookStoreList;
        Article article = articleRepository.findById(id)
                .orElseThrow(() -> new ArticleException(ErrorCode.NOT_FOUND_ARTICLE, id));
        bookStoreList = longs.stream().map(bookStoreId -> bookStoreRepository.findById(bookStoreId)
                .orElseThrow(() -> new ArticleException(ErrorCode.NOT_FOUND_BOOKSTORE, bookStoreId))).toList();
        Object[] objects = article.getArticleBookStoreList().toArray();
        for (Object object : objects) {
            ArticleBookStore articleBookStore = (ArticleBookStore) object;
            articleBookStore.getArticle().removeArticleAndBookStore(articleBookStore);
            articleBookStore.getBookStore().removeArticleBookStore(articleBookStore);
            articleBookStoreRepository.delete(articleBookStore);
        }
        return getArticleResponse(bookStoreList, article);
    }

    @Transactional
    public ArticleWebPageResponse searchArticleList(PageStateRequest pageStateRequest) {
        Pageable pageable = PageRequest.of(pageStateRequest.page() - 1, pageStateRequest.row());
        String search = pageStateRequest.search();
        String articleCategory = pageStateRequest.category();
        ArticleCategory category = ArticleCategory.valueOf(articleCategory);
        String articleStatus = pageStateRequest.status();
        Status status = Status.valueOf(articleStatus);
        Page<ArticleWebResponse> articlePage;
        if (search == null && category == null && status == null) {
            articlePage = articleRepository.findAll(pageable).map(ArticleWebResponse::of);
        } else if (search == null && category == null) {
            articlePage = articleRepository.findAllByStatus(status, pageable).map(ArticleWebResponse::of);
        } else if (search == null && status == null) {
            articlePage = articleRepository.findAllByCategory(category, pageable).map(ArticleWebResponse::of);
        } else if (category == null && status == null) {
            articlePage = articleRepository.findAllByTitleContaining(search, pageable).map(ArticleWebResponse::of);
        } else if (search == null) {
            articlePage = articleRepository.findAllByCategoryAndStatus(category, status, pageable).map(ArticleWebResponse::of);
        } else if (category == null) {
            articlePage = articleRepository.findAllByTitleContainingAndStatus(search, status, pageable).map(ArticleWebResponse::of);
        } else if (status == null) {
            articlePage = articleRepository.findAllByTitleContainingAndCategory(search, category, pageable).map(ArticleWebResponse::of);
        } else {
            articlePage = articleRepository.findAllByTitleContainingAndCategoryAndStatus(search, category, status, pageable).map(ArticleWebResponse::of);
        }

        return ArticleWebPageResponse.of(articlePage);
    }

    @Transactional
    public void removeBookStore(Long id) {
        accountService.isAccountAdmin();
        Article article = articleRepository.findById(id)
                .orElseThrow(() -> new ArticleException(ErrorCode.NOT_FOUND_ARTICLE, id));
        article.softDelete();
    }

    @Transactional
    public Message deleteArticleList(ArticleListRequest list) {
        accountService.isAccountAdmin();

        for (Long id : list.articleIdList()) {
            Article article = articleRepository.findById(id)
                    .orElseThrow(() -> new ArticleException(ErrorCode.NOT_FOUND_ARTICLE, id));
            article.softDelete();
        }
        // 예외처리
        return Message.of("삭제완료");
    }

    @Transactional
    public Message updateArticleStatus(Long id) {
        accountService.isAccountAdmin();
        Article article = articleRepository.findById(id)
                .orElseThrow(() -> new ArticleException(ErrorCode.NOT_FOUND_ARTICLE, id));
        article.updateArticleStatus(article.getStatus() == Status.VISIBLE ? Status.INVISIBLE : Status.VISIBLE);
        return Message.of("상태 변경 완료");
    }
}

File: src/main/kotlin/kr/co/bookand/backend/article/service/ArticleService.kt
package kr.co.bookand.backend.article.service

import kr.co.bookand.backend.account.model.Account
import kr.co.bookand.backend.article.dto.*
import kr.co.bookand.backend.article.model.Article
import kr.co.bookand.backend.article.model.ArticleTag
import kr.co.bookand.backend.article.model.IntroducedBookstore
import kr.co.bookand.backend.article.repository.ArticleRepository
import kr.co.bookand.backend.article.repository.ArticleTagRepository
import kr.co.bookand.backend.article.repository.IntroducedBookstoreRepository
import kr.co.bookand.backend.bookmark.model.BookmarkType
import kr.co.bookand.backend.bookmark.service.BookmarkService
import kr.co.bookand.backend.bookstore.model.Bookstore
import kr.co.bookand.backend.bookstore.dto.BookstoreSimpleResponse
import kr.co.bookand.backend.bookstore.repository.BookstoreRepository
import kr.co.bookand.backend.common.ErrorCode
import kr.co.bookand.backend.common.Status
import kr.co.bookand.backend.common.model.MessageResponse
import kr.co.bookand.backend.common.exception.BookandException
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)

class ArticleService(
    private val articleRepository: ArticleRepository,
    private val bookstoreRepository: BookstoreRepository,
    private val introducedBookstoreRepository: IntroducedBookstoreRepository,
    private val articleTagRepository: ArticleTagRepository,
    private val bookmarkService: BookmarkService
) {
    @Transactional
    fun createArticle(
        currentAccount: Account,
        articleRequest: ArticleRequest
    ): ArticleIdResponse {
        currentAccount.role.checkAdminAndManager()
        duplicateArticle(articleRequest.title)

        val bookStoreList = articleRequest.bookstoreList
        val articleTags = articleRequest.articleTagList
        val article = Article(articleRequest)
        val saveArticle = articleRepository.save(article)

        addBookstoresToArticle(saveArticle, bookStoreList)
        addArticleTags(saveArticle, articleTags)

        return ArticleIdResponse(saveArticle.id)
    }

    @Transactional
    fun updateArticle(
        currentAccount: Account,
        articleId: Long,
        articleRequest: ArticleRequest
    ): ArticleIdResponse {
        currentAccount.role.checkAdminAndManager()

        val article = getArticle(articleId)
        article.updateArticleData(articleRequest)

        removeIntroducedBookstores(article)
        addBookstoresToArticle(article, articleRequest.bookstoreList)
        removeArticleTags(article)
        addArticleTags(article, articleRequest.articleTagList)

        return ArticleIdResponse(article.id)
    }

    fun getArticleList(
        currentAccount: Account,
        pageable: Pageable,
        cursorId: Long?
    ): ArticlePageResponse {
        var nextCursorId = cursorId ?: 0L

        if (cursorId != null && cursorId == 0L) {
            val firstArticle = articleRepository.findFirstByStatusOrderByCreatedAtDesc(Status.VISIBLE)
                ?: throw BookandException(ErrorCode.NOT_FOUND_ARTICLE)
            nextCursorId = firstArticle.id
        }
        val date: String? = if (cursorId == null) null else getArticle(nextCursorId).createdAt.toString()
        val articlePageResponse: Page<ArticleResponse> = articleRepository
            .findAllByStatus(Status.VISIBLE, pageable, nextCursorId, date)
            .map { article ->
                ArticleResponse(
                    id = article.id,
                    title = article.title,
                    subTitle = article.subTitle,
                    mainImage = article.mainImage,
                    category = article.category,
                    content = article.content,
                    writer = article.writer,
                    status = article.status,
                    view = article.viewCount,
                    isBookmark = bookmarkService.checkBookmark(
                        currentAccount.id,
                        article.id,
                        BookmarkType.ARTICLE.name
                    ),
                    articleTagList = article.articleTagList.map { it.tag },
                    visibility = article.visibility,
                    createdDate = article.createdAt.toString(),
                    modifiedDate = article.modifiedAt.toString()
                )
            }

        val totalElements = articleRepository.countAllByStatus(Status.VISIBLE)
        return ArticlePageResponse.of(articlePageResponse, totalElements)
    }

    fun getWebArticleList(currentAccount: Account, pageable: Pageable): ArticleWebPageResponse {
        val articlePage = articleRepository.findAll(pageable)
            .map(::ArticleWebResponse)
        return ArticleWebPageResponse.of(articlePage)

    }

    fun searchArticleList(
        search: String?,
        category: String?,
        status: String?,
        pageable: Pageable
    ): ArticleWebPageResponse {
        val articlePage = articleRepository.findAllBySearch(search, category, status, pageable)
            .map(::ArticleWebResponse)
        return ArticleWebPageResponse.of(articlePage)
    }


    fun getArticleInfo(currentAccount: Account, id: Long): ArticleDetailResponse {
        val article = getArticle(id)

        val bookmarkTypeArticle = BookmarkType.ARTICLE.name
        val bookmarkTypeBookstore = BookmarkType.BOOKSTORE.name
        val accountId = currentAccount.id
        val articleId = article.id

        val bookmarkedArticle = bookmarkService.checkBookmark(accountId, articleId, bookmarkTypeArticle)
        val articleTagList = article.articleTagList.map { it.tag }
        val bookStoreList = article.introducedBookstoreList.map { introducedBookstore ->
            val bookstore = introducedBookstore.bookstore
            val bookstoreId = bookstore.id
            val bookmarkedBookstore = bookmarkService.checkBookmark(accountId, bookstoreId, bookmarkTypeBookstore)

            BookstoreSimpleResponse(
                id = bookstoreId,
                name = bookstore.name,
                introduction = bookstore.introduction,
                mainImage = bookstore.mainImage,
                themeList = bookstore.themeList.map { it.theme.toString() },
                isBookmark = bookmarkedBookstore
            )
        }

        val filter = ArticleFilter(
            deviceOS = article.deviceOSFilter,
            memberId = article.memberIdFilter
        )


        return ArticleDetailResponse(
            id = article.id,
            title = article.title,
            subTitle = article.subTitle,
            mainImage = article.mainImage,
            content = article.content,
            category = article.category,
            writer = article.writer,
            status = article.status,
            view = article.viewCount,
            bookmark = bookmarkedArticle,
            visibility = article.visibility,
            articleTagList = articleTagList,
            bookStoreList = bookStoreList,
            filter = filter,
            createdDate = article.createdAt.toString(),
            modifiedDate = article.modifiedAt.toString()
        )
    }


    private fun addBookstoresToArticle(article: Article, bookstoreList: List<Long>) {
        bookstoreList.map { getBookstore(it) }.toList().forEach { bookstore ->
            val introducedBookstore = IntroducedBookstore(IntroducedBookstoreRequest(article, bookstore))
            val savedIntroducedBookstore = introducedBookstoreRepository.save(introducedBookstore)
            article.updateIntroducedBookstore(savedIntroducedBookstore)
            bookstore.updateIntroducedBookstore(savedIntroducedBookstore)
        }
    }


    private fun addArticleTags(article: Article, tagList: List<String>) {
        tagList.map { ArticleTag(tag = it) }.forEach { articleTag ->
            val savedArticleTag = articleTagRepository.save(articleTag)
            article.updateArticleTag(savedArticleTag)
            savedArticleTag.updateArticle(article)
        }
    }


    @Transactional
    fun updateArticleStatus(articleId: Long): ArticleIdResponse {
        val article = getArticle(articleId)
        val status = article.status
        if (status == Status.VISIBLE) article.updateArticleStatus(Status.INVISIBLE)
        else article.updateArticleStatus(Status.VISIBLE)
        return ArticleIdResponse(article.id)
    }

    fun removeArticle(currentAccount: Account, articleId: Long) {
        currentAccount.role.checkAdminAndManager()
        softDeleteArticle(getArticle(articleId))
    }

    fun removeArticleList(currentAccount: Account, list: ArticleListRequest) {
        currentAccount.role.checkAdminAndManager()
        list.articleIdList.forEach { articleId ->
            softDeleteArticle(getArticle(articleId))
        }
    }

    fun softDeleteArticle(article: Article){
        removeIntroducedBookstores(article)
        removeArticleTags(article)
        article.softDelete()
    }

    fun duplicateArticle(title: String) {
        if (isDuplicateArticle(title)) {
            throw BookandException(ErrorCode.DUPLICATE_ARTICLE)
        }
    }

    fun getBookstore(bookstoreId: Long): Bookstore {
        return bookstoreRepository.findById(bookstoreId)
            .orElseThrow { BookandException(ErrorCode.NOT_FOUND_BOOKSTORE) }
    }

    fun getArticle(articleId: Long): Article {
        return articleRepository.findById(articleId)
            .orElseThrow { BookandException(ErrorCode.NOT_FOUND_ARTICLE) }
    }

    private fun removeIntroducedBookstores(article: Article) {
        article.introducedBookstoreList.toList().forEach { introducedBookstore ->
            introducedBookstoreRepository.delete(introducedBookstore)
            introducedBookstore.bookstore.removeIntroducedBookstore(introducedBookstore)
            article.removeIntroducedBookstore(introducedBookstore)
        }
    }

    private fun removeArticleTags(article: Article) {
        article.articleTagList.toList().forEach { articleTag ->
            articleTagRepository.delete(articleTag)
            article.removeArticleTag(articleTag)
        }
    }

    private fun isDuplicateArticle(title: String): Boolean {
        return articleRepository.existsByTitle(title)
    }
}